### PR TITLE
Add new TiCo Model.

### PR DESCRIPTION
**Pull request by Ra1Nik**
- Implement TiCo Loss and Projection Head
- Add Example
- Add Unit Test
- Update Docs

### CODE REVIEW

**Comment by Ra1Nik** [comment]
Renderings of the docs:

[lightly.loss — lightly 1.2.40 documentation.pdf](https://github.com/lightly-ai/lightly/files/10276538/lightly.loss.lightly.1.2.40.documentation.pdf)
[TiCo — lightly 1.2.40 documentation.pdf](https://github.com/lightly-ai/lightly/files/10276539/TiCo.lightly.1.2.40.documentation.pdf)

File: examples/pytorch/tico.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+criterion = TiCoLoss(
+    beta_param = 0.9,
+    ro_param = 20.0,
+)
+
+optimizer = torch.optim.SGD(model.parameters(), lr=0.06)
+
+C_prev = Variable(torch.zeros(256, 256), requires_grad=True).to(device)
+C_prev = C_prev.detach()
```
Comment (guarin) [inline examples/pytorch/tico.py:74]: Can we put `C` into `TiCoLoss` instead of passing it around?

I would propose to use the following definition for the forward method in the loss:
```
def forward(self, z_a: torch.Tensor, z_b: torch.Tensor, update_covariance_matrix: bool = True) -> torch.Tensor:
# only update C if update_covariance_matrix == True
```

File: examples/pytorch/tico.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    batch_size=8,
+    collate_fn=collate_fn,
+    shuffle=True,
+    drop_last=True,
+    num_workers=8,
+)
+
+criterion = TiCoLoss(
+    beta_param = 0.9,
+    ro_param = 20.0,
```
Comment (guarin) [inline examples/pytorch/tico.py:68]: No need to set `beta_param` and `ro_param` as they are already the default values.

File: lightly/loss/tico_loss.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+from lightly.utils.dist import gather
+
+class TiCoLoss(torch.nn.Module):
+    """Implementation of the Tico Loss from Tico[0] paper.
+    This implementation takes inspiration from the code published 
+    by sayannag using Lightly. [1]
+
+    [0] Jiachen Zhu et. al, 2022, Tico... https://arxiv.org/abs/2206.10698
+    [1] https://github.com/sayannag/TiCo-pytorch
```
Comment (guarin) [inline lightly/loss/tico_loss.py:14]: Could you format this docstring in the same format as the one for `DINOLoss`?

File: lightly/loss/tico_loss.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        >>>
+        >>> # calculate loss
+        >>> loss = loss_fn(out0, out1)
+    """
+
+    def __init__(
+        self,
+        beta_param: float = 0.9,
+        ro_param: float = 20.0,
+        gather_distributed : bool = False,
```
Comment (guarin) [inline lightly/loss/tico_loss.py:36]: -> `gather_distributed: bool = False,`

(remove whitespace)

File: lightly/loss/tico_loss.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        >>> out0, out1 = model(t0, t1)
+        >>>
+        >>> # calculate loss
+        >>> loss = loss_fn(out0, out1)
+    """
+
+    def __init__(
+        self,
+        beta_param: float = 0.9,
+        ro_param: float = 20.0,
```
Comment (guarin) [inline lightly/loss/tico_loss.py:35]: `beta_param` -> `beta`
`ro_param` -> `rho`

File: lightly/loss/tico_loss.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        gather_distributed : bool = False,
+    ):
+        """Lambda, mu and nu params configuration with default value like in [0]
+        Args:
```
Comment (guarin) [inline lightly/loss/tico_loss.py:39]: Where do lambda, mu, and nu come from? I cannot find them in the paper.

Could you also add a newline before `Args:`?

File: lightly/loss/tico_loss.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        # gather all batches
+        if self.gather_distributed and dist.is_initialized():
+            world_size = dist.get_world_size()
+            if world_size > 1:
+                z_a = torch.cat(gather(z_a), dim=0)
+                z_b = torch.cat(gather(z_b), dim=0)
+
+        # normalize repr. along the batch dimension
+        z_a = z_a - z_a.mean(0) # NxD
+        z_b = z_b - z_b.mean(0) # NxD
```
Comment (guarin) [inline lightly/loss/tico_loss.py:69]: Could not find this in the paper. Are you sure we have to do this?

File: lightly/loss/tico_loss.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+        # normalize image
+        z_a = torch.nn.functional.normalize(z_a, dim = -1)
+        z_b = torch.nn.functional.normalize(z_b, dim = -1)
```
Comment (guarin) [inline lightly/loss/tico_loss.py:73]: I believe dim should be `1` instead of `-1`. Eventually, we have to flatten the representations first:
`z_a.flatten(start_dim=1)`.

File: lightly/loss/tico_loss.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        
+        # compute loss
+        B = torch.mm(z_a.T, z_a)/z_a.shape[0]        
+        C = self.beta_param * C + (1 - self.beta_param) * B
+        loss = 1 - (z_a * z_b).sum(dim=1).mean() + self.ro_param * (torch.mm(z_a, C) * z_a).sum(dim=1).mean()
```
Comment (guarin) [inline lightly/loss/tico_loss.py:78]: Why `1 - (z_a, * z_b )...`? I can only find ` - (z_a * z_b) ...` in the paper.

File: lightly/models/utils.py [hunk 10]
```diff
@@ -461,3 +462,31 @@ def random_token_mask(
...
+
+def schedule_momentum(
+    iteration: int, 
+    max_iter: int, 
+    m: float = 0.99
+) -> float:
+
+    """
+    Add sinusoidal noise to momentum value based on iteration number. Common
+    training tecnique in SSL
```
Comment (guarin) [inline lightly/models/utils.py:474]: tecnique -> technique

Also, missing `.` at end of sentence 😉

File: lightly/models/utils.py [hunk 11]
```diff
@@ -461,3 +462,31 @@ def random_token_mask(
...
+
+    Args:
+        iteration:
+            Size of the token batch for which to generate masks.
+            Should be (batch_size, sequence_length).
+        max_iter:
+            Percentage of tokens to mask.
+        m:
+            If False the class token is never masked. If True the class token
+            might be masked.
```
Comment (guarin) [inline lightly/models/utils.py:484]: Why does this mention mask tokens? Could you update the docstrings?

File: tests/loss/test_TicoLoss.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
+import unittest
+import torch
+
+from lightly.loss.tico_loss import TiCoLoss
+torch.manual_seed(0)
+class TestVICRegLoss(unittest.TestCase):
```
Comment (guarin) [inline tests/loss/test_TicoLoss.py:6]: `TestTiCoLoss`

File: tests/loss/test_TicoLoss.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
+import unittest
+import torch
+
+from lightly.loss.tico_loss import TiCoLoss
+torch.manual_seed(0)
```
Comment (guarin) [inline tests/loss/test_TicoLoss.py:5]: Please add whitespace after imports and move `torch.manual_seed(0)` into each test method.

**Review by guarin — CHANGES_REQUESTED** [review]
Awesome stuff!

Left a bunch of comments, will have a second look tomorrow :)

File: lightly/loss/tico_loss.py (hunk 9 above)
Comment (Ra1Nik) [reply lightly/loss/tico_loss.py:78]: Schermata 2022-12-22 alle 10 00 34

Equation (4)

Comment (guarin) [reply lightly/loss/tico_loss.py:78]: Interesting, the pseudocode doesn't have the `1 -`:
Screen Shot 2022-12-22 at 10 07 12

Comment (Ra1Nik) [reply lightly/loss/tico_loss.py:78]: Also the code from the implementation I followed had no 1-. But then I saw that the loss started being negative and spotted the problem.

Comment (guarin) [reply lightly/loss/tico_loss.py:78]: I think you can drop the 1 as it is just a constant added to the loss which should not change training.

Comment (Ra1Nik) [reply lightly/loss/tico_loss.py:78]: Yes, but isn't it good to always have a positive loss?

Comment (Ra1Nik) [reply lightly/loss/tico_loss.py:78]: I would personally let it there as I think that it is not really affecting the computation

File: lightly/models/utils.py [hunk 14]
```diff
@@ -471,20 +471,18 @@ def schedule_momentum(
...
-            Size of the token batch for which to generate masks.
-            Should be (batch_size, sequence_length).
+            Current epoch number.
         max_iter:
-            Percentage of tokens to mask.
+            Total number of epochs.
         m:
-            If False the class token is never masked. If True the class token
-            might be masked.
+            Starting momentum value.
```
Comment (guarin) [inline lightly/models/utils.py:482]: I would propose to make this method more general and rename it to `cosine_schedule` with the following parameters:
```
step (instead of iteration)
max_steps (instead of max_iter)
start_value (instead of m)
end_value (new)
```
The method would then interpolate between `start_value` and `end_value` with a cosine decay.

File: lightly/loss/tico_loss.py (hunk 8 above)
Comment (guarin) [reply lightly/loss/tico_loss.py]: Did you verify that `z_a` has only 2 dimensions?

**Review by guarin — CHANGES_REQUESTED** [review]
Thanks a lot for the changes :)

Made a second sweep and left some comments.

File: lightly/loss/tico_loss.py (hunk 8 above)
Comment (Ra1Nik) [reply lightly/loss/tico_loss.py]: Yes, doing a print. Should I also do it like an assertion?

File: lightly/loss/tico_loss.py [hunk 15]
```diff
@@ -31,31 +47,39 @@ class TiCoLoss(torch.nn.Module):
 
     def __init__(
         self,
-        beta_param: float = 0.9,
-        ro_param: float = 20.0,
-        gather_distributed : bool = False,
+        beta: float = 0.9,
+        rho: float = 20.0,
+        C: torch.Tensor = Variable(torch.zeros(256, 256), requires_grad=True),
```
[the remark this answers is not in the document]
Comment (Ra1Nik) [reply lightly/loss/tico_loss.py:52]: Actually, letting C as an argument might be useful because if someone wants to change the last layer dimension can overwrite C as Variable(torch.zeros(32, 32), requires_grad=True) for example.

File: lightly/models/utils.py (hunk 14 above)
Comment (Ra1Nik) [reply lightly/models/utils.py:482]: What would be the purpose of the end_value? I mean, if you put max steps, it is already an "end value" in the sense that you compute the end value with the step=max_steps

File: lightly/loss/tico_loss.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/loss/tico_loss.py:52]: My proposal below would do this automatically: https://github.com/lightly-ai/lightly/pull/1016#discussion_r1055286323

File: lightly/models/utils.py (hunk 14 above)
Comment (guarin) [reply lightly/models/utils.py:482]: Let's say you want cosine decay from 0.5 to 0.1 over 100 epochs. Then you would set:
```
max_steps = 100
start_value = 0.5
end_value = 0.1
```

File: lightly/models/utils.py [hunk 16]
```diff
@@ -463,28 +463,33 @@ def random_token_mask(
...
+        end_value:
+            Target momentum value.
             
     Returns:
         New momentum value to be used with update_momentum.
 
     """
-    
-    momentum = m + (1 - m)*np.sin((np.pi / 2)* iteration / (max_iter - 1))
+    if (max_steps == 1):
```
Comment (guarin) [inline lightly/models/utils.py:491]: Could you raise `ValueError` for the cases: `step < 0`, or `max_steps < 0`, or `step >= max_steps`.

File: lightly/models/utils.py [hunk 17]
```diff
@@ -463,28 +463,33 @@ def random_token_mask(
 
     return idx_keep, idx_mask
 
-def schedule_momentum(
-    iteration: int, 
-    max_iter: int, 
-    m: float = 0.99
+def cosine_decay_schedule(
+    step: int,
```
Comment (guarin) [inline lightly/models/utils.py:467]: Can we just call it `cosine_schedule`? Decay implies that the value decreases but it also supports increasing values.

File: lightly/models/utils.py [hunk 18]
```diff
@@ -463,28 +463,33 @@ def random_token_mask(
...
 
-def schedule_momentum(
-    iteration: int, 
-    max_iter: int, 
-    m: float = 0.99
+def cosine_decay_schedule(
+    step: int, 
+    max_steps: int, 
+    start_value: float = 0.996,
+    end_value: float = 1
```
Comment (guarin) [inline lightly/models/utils.py:470]: I wouldn't add default values here as the function is very general. It is, for example, also often used for learning rate schedules.

File: lightly/models/utils.py [hunk 19]
```diff
@@ -463,28 +463,33 @@ def random_token_mask(
...
 ) -> float:
 
     """
-    Add sinusoidal noise to momentum value based on iteration number. Common
-    training technique in SSL.
+    Use cosine decay to gradually modify start_value to reach target end_value during training.
+    Introduced in BYOL (https://arxiv.org/pdf/2006.07733.pdf) to update the momentum.
```
Comment (guarin) [inline lightly/models/utils.py:475]: No need to cite anything here. Cosine schedules are common and existed before BYOL :)

File: lightly/loss/tico_loss.py [hunk 20]
```diff
@@ -92,13 +85,19 @@ def forward(self, z_a: torch.Tensor, z_b: torch.Tensor, update_covariance_matrix
         z_a = torch.nn.functional.normalize(z_a, dim = 1)
         z_b = torch.nn.functional.normalize(z_b, dim = 1)
         
+        # compute ausiliary matrix B
```
Comment (guarin) [inline lightly/loss/tico_loss.py:88]: ausiliary -> auxiliary

**Review by guarin — CHANGES_REQUESTED** [review]
Nice, I think we can merge this soon :)

File: examples/pytorch_lightning/tico.py [hunk 21]
```diff
@@ -43,7 +42,7 @@ def forward_momentum(self, x):
 
     def training_step(self, batch, batch_idx):
         (x0, x1), _, _ = batch
-        momentum = cosine_decay_schedule(batch_idx, 10)
+        momentum = cosine_schedule(batch_idx, 10, 0.996, 1)
```
Comment (guarin) [inline examples/pytorch_lightning/tico.py:45]: I believe we have to use current epoch instead of batch idx here:
```
momentum = cosine_schedule(self.current_epoch, 10, 0.996, 1)
```

File: examples/pytorch_lightning_distributed/tico.py [hunk 22]
```diff
@@ -43,7 +42,7 @@ def forward_momentum(self, x):
 
     def training_step(self, batch, batch_idx):
         (x0, x1), _, _ = batch
-        momentum = cosine_decay_schedule(batch_idx, 10)
+        momentum = cosine_schedule(batch_idx, 10, 0.996, 1)
```
Comment (guarin) [inline examples/pytorch_lightning_distributed/tico.py:45]: I guess this should also use self.current_epoch instead of batch_idx

File: lightly/models/utils.py [hunk 23]
```diff
@@ -488,6 +487,10 @@ def cosine_decay_schedule(
         New momentum value to be used with update_momentum.
```
Comment (guarin) [inline lightly/models/utils.py:488]: Could you remove the references to "momentum" in the docstring?

File: lightly/models/utils.py [hunk 24]
```diff
@@ -488,6 +487,10 @@ def cosine_decay_schedule(
...
 
     """
+    assert step >= 0, "Current number can't be negative"
+    assert max_steps > 0, "Total epochs must be > 0"
+    assert step <= max_steps, f"Current step is bigger than total number of steps. Found {step} > {max_steps}."
```
Comment (guarin) [inline lightly/models/utils.py:492]: I would prefer `ValueError` instead of `assert` here.

Comment (guarin) [inline lightly/models/utils.py:492]: Are you sure that it is `step <= max_steps` and not `step < max_steps`? Because further down we have: `step / (max_steps - 1)`

**Review by guarin — CHANGES_REQUESTED** [review]

File: lightly/models/utils.py (hunk 24 above)
Comment (Ra1Nik) [reply lightly/models/utils.py:492]: Yes, as I cut out the max_steps = 1 case with the if clause (and "step" starts from 0) and if I have step = max_steps = 1  is a case that I would like to keep

Comment (guarin) [reply lightly/models/utils.py]: Yes but the calculation doesn't work if we have for example, step=max_steps=4. Then the formula does: `step / (max_steps - 1) = 4 / 3` which is `>1` and we want the schedule to only go to 1 no?

Comment (Ra1Nik) [reply lightly/models/utils.py]: Oh yes sorry, step will never be = max step because as I wrote "step" starts from 0. Sorry

File: lightly/models/utils.py (hunk 23 above)
Comment (guarin) [reply lightly/models/utils.py:488]: There are still references about momentum in the docstring.

File: lightly/models/utils.py [hunk 25]
```diff
@@ -484,12 +484,15 @@ def cosine_schedule(
...
     Returns:
-        New momentum value to be used with update_momentum.
+        Cosine decay value.
 
     """
-    assert step >= 0, "Current number can't be negative"
-    assert max_steps > 0, "Total epochs must be > 0"
-    assert step <= max_steps, f"Current step is bigger than total number of steps. Found {step} > {max_steps}."
+    if step < 0:
+        raise ValueError("Current epoch number can't be negative")
```
Comment (guarin) [inline lightly/models/utils.py:491]: Should be "step" instead of "epoch"

Comment (guarin) [reply lightly/models/utils.py:491]: Also in the other errors.

File: lightly/models/utils.py [hunk 26]
```diff
@@ -484,12 +484,15 @@ def cosine_schedule(
...
+    if max_steps < 1:
+        raise ValueError("Total epochs number must be >= 1")
+    if step >= max_steps:
+        raise ValueError(f"Current step is bigger than total number of steps. Found {step} > {max_steps}.")
```
Comment (guarin) [inline lightly/models/utils.py:495]: The check and the error message do not say the same `step >= max_steps` and `step > max_steps`. 

The error message might be more helpful if it said something along the lines: `The current step must be smaller than max_steps but found step {step} and max_steps {max_steps}.`

**Review by guarin — CHANGES_REQUESTED** [review]

**Review by guarin — APPROVED** [review]
Awesome!
